<a href="https://colab.research.google.com/github/Kbcity1/checkpoint/blob/master/%20Unsupervised%20Association%20Rules%20Checkpoint.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
"""
Instructions

toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],

        ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],

        ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],

        ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],

      ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

Run the apriori algorithm on the provided toy_dataset. Interpret the results.
Try to explore the checkpoint dataset using Pandas and Plotly.
Run the apriori algorithm on checkpoint dataset. Interpret the results and suggest a clear business plan to the supermarket owners based on your findings.

""""

In [1]:
#Importing libraries
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

# Provided toy dataset
toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],
               ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],
               ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],
               ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],
               ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

In [2]:
# Transform the dataset into a format suitable for the apriori algorithm
te = TransactionEncoder()
te_ary = te.fit(toy_dataset).transform(toy_dataset) #Apply one-hot-encoding on our dataset
df = pd.DataFrame(te_ary, columns=te.columns_) #Creating a nw Dataframe from our numpy array

df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Dress  Earrings    Hat  Heels  Necklace  Pants  Sandals  Scarf  Shirt  \
0  False     False   True  False     False   True    False   True  False   
1  False     False   True  False     False   True    False  False  False   
2   True     False  False   True     False   True     True   True  False   
3   True      True   True   True      True  False    False   True  False   
4  False      True  False  False     False   True    False   True   True   

   Skirt  Sneakers  Sunglasses  
0   True      True       False  
1   True      True        True  
2  False     False       False  
3  False     False       False  
4   True     False       False

In [3]:
# Transform the dataset into a format suitable for the apriori algorithm
te = TransactionEncoder()
te_ary = te.fit(toy_dataset).transform(toy_dataset) #Apply one-hot-encoding on our dataset
df = pd.DataFrame(te_ary, columns=te.columns_) #Creating a nw Dataframe from our numpy array

df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Dress  Earrings    Hat  Heels  Necklace  Pants  Sandals  Scarf  Shirt  \
0  False     False   True  False     False   True    False   True  False   
1  False     False   True  False     False   True    False  False  False   
2   True     False  False   True     False   True     True   True  False   
3   True      True   True   True      True  False    False   True  False   
4  False      True  False  False     False   True    False   True   True   

   Skirt  Sneakers  Sunglasses  
0   True      True       False  
1   True      True        True  
2  False     False       False  
3  False     False       False  
4   True     False       False

In [4]:
#Try to explore the checkpoint dataset using Pandas and Plotly.
import plotly.express as px

# Calculate item frequencies
item_frequencies = df.sum()  # Calculate sum of each column (item)

# Create a DataFrame for plotting
df_plot = pd.DataFrame({'Item': item_frequencies.index, 'Frequency': item_frequencies.values})

# Bar plot of item frequencies
fig = px.bar(df_plot, x='Item', y='Frequency', title='Frequency of Items in Toy Dataset')
fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [5]:
# Apply the apriori algorithm
frequent_itemsets = apriori(df, min_support=0.4, use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



support                       itemsets
0       0.4                        (Dress)
1       0.4                     (Earrings)
2       0.6                          (Hat)
3       0.4                        (Heels)
4       0.8                        (Pants)
5       0.8                        (Scarf)
6       0.6                        (Skirt)
7       0.4                     (Sneakers)
8       0.4                 (Dress, Heels)
9       0.4                 (Dress, Scarf)
10      0.4              (Earrings, Scarf)
11      0.4                   (Pants, Hat)
12      0.4                   (Hat, Scarf)
13      0.4                   (Skirt, Hat)
14      0.4                (Sneakers, Hat)
15      0.4                 (Heels, Scarf)
16      0.6                 (Pants, Scarf)
17      0.6                 (Pants, Skirt)
18      0.4              (Pants, Sneakers)
19      0.4                 (Skirt, Scarf)
20      0.4              (Sneakers, Skirt)
21      0.4          (Dress, Heels, Scarf)
22      0.4            (Pants, Skirt, Hat)
23      0.4         (Pants, Sneakers, Hat)
24      0.4         (Sneakers, Skirt, Hat)
25      0.4          (Pants, Skirt, Scarf)
26      0.4       (Pants, Sneakers, Skirt)
27      0.4  (Pants, Sneakers, Skirt, Hat)

INTERPRET THE RESULT

*   Pants and scarves are the most frequently purchased items, appearing in 80% of transactions.

*   The combination of hat, sneakers, pants, and skirt appears in 40% of transactions.

*  Pants and scarf are frequently bought together, appearing in 60% of transactions.

In [7]:
#to know the number of itemsset
num_itemsets = len(frequent_itemsets)
print(f"Number of frequent itemsets: {num_itemsets}")

Number of frequent itemsets: 28


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [8]:
# Generate the association rules
rules = association_rules(frequent_itemsets, metric="confidence", num_itemsets = 28, min_threshold=0.6)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [9]:
print("Frequent Itemsets:\n", frequent_itemsets)

Frequent Itemsets:
     support                       itemsets
0       0.4                        (Dress)
1       0.4                     (Earrings)
2       0.6                          (Hat)
3       0.4                        (Heels)
4       0.8                        (Pants)
5       0.8                        (Scarf)
6       0.6                        (Skirt)
7       0.4                     (Sneakers)
8       0.4                 (Dress, Heels)
9       0.4                 (Dress, Scarf)
10      0.4              (Earrings, Scarf)
11      0.4                   (Pants, Hat)
12      0.4                   (Hat, Scarf)
13      0.4                   (Skirt, Hat)
14      0.4                (Sneakers, Hat)
15      0.4                 (Heels, Scarf)
16      0.6                 (Pants, Scarf)
17      0.6                 (Pants, Skirt)
18      0.4              (Pants, Sneakers)
19      0.4                 (Skirt, Scarf)
20      0.4              (Sneakers, Skirt)
21      0.4          (Dress, Heels

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [10]:
print("\nAssociation Rules:\n", rules)


Association Rules:
         antecedents               consequents  antecedent support  \
0           (Dress)                   (Heels)                 0.4   
1           (Heels)                   (Dress)                 0.4   
2           (Dress)                   (Scarf)                 0.4   
3        (Earrings)                   (Scarf)                 0.4   
4             (Hat)                   (Pants)                 0.6   
..              ...                       ...                 ...   
57  (Sneakers, Hat)            (Pants, Skirt)                 0.4   
58     (Skirt, Hat)         (Pants, Sneakers)                 0.4   
59       (Sneakers)       (Pants, Skirt, Hat)                 0.4   
60          (Skirt)    (Pants, Hat, Sneakers)                 0.6   
61            (Hat)  (Pants, Skirt, Sneakers)                 0.6   

    consequent support  support  confidence      lift  representativity  \
0                  0.4      0.4    1.000000  2.500000               1.0   


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



Interpretation of Association Rules

*   Dress and Heels: Customers who buy dresses also buy heels with 100% confidence, indicating a strong correlation (lift of 2.5).
*   Hat, Sneakers, Pants, and Skirt: The combination of these items frequently occurs together, indicating a strong relationship (lift of 2.5).
*   Dress and Scarf: Dresses are often purchased with scarves, showing a confidence of 100% but with a lower lift (1.25).
*   Earrings and Scarf: Earrings are frequently bought with scarves, also with a confidence of 100% and a lift of 1.25.


Business Plan Recommendations


*   Create promotional bundles for dresses and heels. Given the strong association, customers are likely to appreciate discounts or special offers on these combinations.

*   Since dresses are frequently purchased with scarves, consider offering a discount on scarves with the purchase of a dress. This can encourage customers to buy both items.


In [12]:
import pandas as pd

# Assuming the dataset is in a CSV file named 'customer_purchases.csv'
df = pd.read_csv('/content/Market_Basket_Optimisation.csv', header =None)

# Display the first few rows
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [14]:
import pandas as pd
import plotly.express as plt

Items = {}
# Access the first column (index 0) since there are no headers
for item in df.iloc[:, 0]:
    if item in Items:
        Items[item] = Items[item] + 1
    else:
        Items[item] = 1

keys = []
vals = []
for i, k in Items.items():
    if k > 30:
        keys.append(i)
        vals.append(k)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [15]:
# Create a DataFrame for Plotly Express
df_plot = pd.DataFrame({'Items': keys, 'Transactions': vals})

# Use the 'text' argument to display labels on the bars
fig = plt.bar(df_plot, x='Items', y='Transactions', text='Transactions')

fig.update_layout(
    xaxis_title='Items Sold',
    yaxis_title='Number of Transactions',
    title='Items frequency',
    xaxis_tickangle=90,
    width=1000,  # Adjust width as needed
    height=600   # Adjust height as needed
)

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [16]:
# Display general information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
 1   1       5747 non-null   object
 2   2       4389 non-null   object
 3   3       3345 non-null   object
 4   4       2529 non-null   object
 5   5       1864 non-null   object
 6   6       1369 non-null   object
 7   7       981 non-null    object
 8   8       654 non-null    object
 9   9       395 non-null    object
 10  10      256 non-null    object
 11  11      154 non-null    object
 12  12      87 non-null     object
 13  13      47 non-null     object
 14  14      25 non-null     object
 15  15      8 non-null      object
 16  16      4 non-null      object
 17  17      4 non-null      object
 18  18      3 non-null      object
 19  19      1 non-null      object
dtypes: object(20)
memory usage: 1.1+ MB


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [17]:
import numpy as np

# Convert DataFrame to list of lists
transactions = df.applymap(str).values.tolist()

# Remove 'nan' values from the transactions
transactions = [[item for item in transaction if item != 'nan'] for transaction in transactions]

# Display the first few transactions
transactions[:5]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.

<ipython-input-17-3ca9e86f8eaa>:4: FutureWarning:

DataFrame.applymap has been deprecated. Use DataFrame.map instead.



[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea']]

In [18]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

# Convert dataset into a one-hot encoded DataFrame
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df_onehot = pd.DataFrame(te_ary, columns=te.columns_)

df_onehot.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

In [19]:
# Run Apriori algorithm
frequent_itemsets = apriori(df_onehot, min_support=0.01, use_colnames=True)  # Adjust min_support as needed

frequent_itemsets.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



support          itemsets
0  0.020397         (almonds)
1  0.033329         (avocado)
2  0.010799  (barbecue sauce)
3  0.014265       (black tea)
4  0.011465      (body spray)

In [20]:
num_itemsets = len(frequent_itemsets)
print(f"Number of frequent itemsets: {num_itemsets}")

Number of frequent itemsets: 257


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [21]:
# Generate association rules
rules = association_rules(frequent_itemsets, metric="confidence", num_itemsets = 257, min_threshold=0.5)

print("Frequent Itemsets:\n")
frequent_itemsets.head()

Frequent Itemsets:



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



support          itemsets
0  0.020397         (almonds)
1  0.033329         (avocado)
2  0.010799  (barbecue sauce)
3  0.014265       (black tea)
4  0.011465      (body spray)

In [22]:
print("\nAssociation Rules:\n")
rules.head()


Association Rules:



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



antecedents      consequents  antecedent support  \
0  (ground beef, eggs)  (mineral water)            0.019997   
1  (ground beef, milk)  (mineral water)            0.021997   

   consequent support   support  confidence      lift  representativity  \
0            0.238368  0.010132    0.506667  2.125563               1.0   
1            0.238368  0.011065    0.503030  2.110308               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.005365    1.543848       0.540342  0.040816   0.352268    0.274586  
1  0.005822    1.532552       0.537969  0.044385   0.347493    0.274725

Interpretation of the Results
Association Rules:



1.   **Rule 1**: (eggs, ground beef) → (mineral water)


*   **Antecedent Support**: 0.019997 (approx. 2% of transactions include both eggs and ground beef)

*   **Consequent Support**: 0.238368 (mineral water appears in approx. 23.84% of transactions)

*   **Support**: 0.010132 (approx. 1.01% of transactions include eggs, ground beef, and mineral water)

*   **Confidence**: 0.506667 (50.67% of transactions with eggs and ground beef also include mineral water)

*   **Lift**: 2.125563 (mineral water is 2.13 times more likely to be bought with eggs and ground beef than by itself)



2.   **Rule 2**: (milk, ground beef) → (mineral water)


*   **Antecedent Support**: 0.021997 (approx. 2.20% of transactions include both milk and ground beef)

*   **Consequent Support**: 0.238368 (mineral water appears in approx. 23.84% of transactions)

*   **Support**: 0.011065 (approx. 1.11% of transactions include milk, ground beef, and mineral water)

*   **Confidence**: 0.503030 (50.30% of transactions with milk and ground beef also include mineral water)

*   **Lift**: 2.110308 (mineral water is 2.11 times more likely to be bought with milk and ground beef than by itself)


**Business Plan Suggestions**
Based on these findings, here are some actionable business strategies for the supermarket:


1.   **Product Bundling:**


*   **Bundle Suggestion 1:** Create bundles or promotional offers combining eggs, ground beef, and mineral water. Given the strong association (50.67% confidence), promoting these items together can drive sales.

*   **Bundle Suggestion 2:** Similarly, bundle milk, ground beef, and mineral water. With a 50.30% confidence, this combination is also highly popular among customers.

2.   **Store Layout Optimization:**

*   Place mineral water near the meat section, especially around eggs and ground beef, as customers who purchase these items are likely to buy mineral water as well. This can increase the visibility of mineral water and potentially boost its sales.

*   Position frequently bought items like avocado and almonds at eye level or in high-traffic areas to capitalize on their popularity.

3.   **Targeted Promotions:**

*   Offer discounts on mineral water for customers who purchase eggs and ground beef or milk and ground beef. This can encourage customers to add mineral water to their shopping cart, leveraging the strong association rules.

*   Use loyalty programs to provide personalized discounts on items like avocado, black tea, and body spray, which have a moderate frequency but can be promoted further.








